// rrx/dist/createRouter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createHistory = require('./createHistory');

var _createHistory2 = _interopRequireDefault(_createHistory);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var createRouter = function createRouter(Component) {
  var RouterComponent = function (_React$Component) {
    _inherits(RouterComponent, _React$Component);

    function RouterComponent(_ref) {
      var routes = _ref.routes,
          pathname = _ref.pathname;

      _classCallCheck(this, RouterComponent);

      var _this = _possibleConstructorReturn(this, (RouterComponent.__proto__ || Object.getPrototypeOf(RouterComponent)).call(this));

      _this.state = {
        history: {},
        location: {}
      };
      _this.handleHistoryChange = _this.handleHistoryChange.bind(_this);
      _this.unlisten = function () {};
      return _this;
    }

    _createClass(RouterComponent, [{
      key: 'getChildContext',
      value: function getChildContext() {
        return this.state;
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        var history = (0, _createHistory2.default)(this.props.options);
        var location = history.location;

        this.setState({ history: history, location: location });
        this.unlisten = history.listen(this.handleHistoryChange);
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        this.unlisten();
      }
    }, {
      key: 'handleHistoryChange',
      value: function handleHistoryChange(location) {
        this.setState({ location: location });
      }
    }, {
      key: 'render',
      value: function render() {
        return _react2.default.createElement(Component, _extends({}, this.props, this.state));
      }
    }]);

    return RouterComponent;
  }(_react2.default.Component);

  RouterComponent.childContextTypes = {
    history: _propTypes2.default.object,
    location: _propTypes2.default.object
  };

  RouterComponent.propTypes = {
    options: _propTypes2.default.object
  };

  return RouterComponent;
};

exports.default = createRouter;
